src/bsp/platform/esp32/xi_bsp_io_fs_esp32.c

```diff
@@ -0,0 +1,307 @@
+#define ESP32_FS_BASE_PATH "/spiflash"
+
+#define XI_BSP_IO_FS_CHECK_CND( cnd, e, s )                                              \
+    if ( ( cnd ) )                                                                       \
+    {                                                                                    \
+        ( s ) = ( e );                                                                   \
+        perror( "Xi FS BSP errno" );                                                     \
+        goto err_handling;                                                               \
+    }
+
+/* These functions are implemented in the application layer to report status */
+extern void esp32_xibsp_notify_update_started( const char* filename, size_t file_size );
+extern void esp32_xibsp_notify_chunk_written( size_t chunk_size, size_t offset );
+
+/* The size of the buffer to be used for reads. */
+const size_t xi_bsp_io_fs_buffer_size = 1024;
+/* ESP32 Over The Air FW updates API handle */
+static esp_ota_handle_t open_firmware_bin_handle = 0;
+
+/* translates bsp errno errors to the xi_bsp_io_fs_state_t values */
+xi_bsp_io_fs_state_t xi_bsp_io_fs_posix_errno_2_xi_bsp_io_fs_state( int errno_value )
+{
+    xi_bsp_io_fs_state_t ret = XI_BSP_IO_FS_STATE_OK;
+
+    switch ( errno_value )
+    {
+        case EBADF:
+        case EACCES:
+        case EFAULT:
+        case ENOENT:
+        case ENOTDIR:
+            ret = XI_BSP_IO_FS_RESOURCE_NOT_AVAILABLE;
+            break;
+        case ELOOP:
+        case EOVERFLOW:
+            ret = XI_BSP_IO_FS_ERROR;
+            break;
+        case EINVAL:
+            ret = XI_BSP_IO_FS_INVALID_PARAMETER;
+            break;
+        case ENAMETOOLONG:
+            ret = XI_BSP_IO_FS_INVALID_PARAMETER;
+            break;
+        case ENOMEM:
+            ret = XI_BSP_IO_FS_OUT_OF_MEMORY;
+            break;
+        default:
+            ret = XI_BSP_IO_FS_ERROR;
+    }
+
+    return ret;
+}
+
+xi_bsp_io_fs_state_t
+xi_bsp_io_fs_open( const char* const resource_name,
+                   const size_t size,
+                   const xi_bsp_io_fs_open_flags_t open_flags,
+                   xi_bsp_io_fs_resource_handle_t* resource_handle_out )
+{
+    char filepath[XI_ESP32_MAX_FILENAME_LEN] = "";
+    xi_bsp_io_fs_state_t ret                 = XI_BSP_IO_FS_STATE_OK;
+    esp_err_t esp_retv                       = ESP_OK;
+    FILE* fp                                 = NULL;
+
+    if ( NULL == resource_name || NULL == resource_handle_out )
+    {
+        return XI_BSP_IO_FS_INVALID_PARAMETER;
+    }
+
+    if ( xi_bsp_fwu_is_this_firmware( resource_name ) )
+    {
+        const esp_partition_t* next_partition = esp_ota_get_next_update_partition( NULL );
+
+        if ( NULL == next_partition )
+        {
+            xi_bsp_debug_logger( "Failed to get next OTA update partition" );
+            return XI_BSP_IO_FS_OPEN_ERROR;
+        }
+
+        esp_retv = esp_ota_begin( next_partition, size,
```

Comment:
wow, size is needed here too. POSIX doesn't, CC3200 does need the size here in the open file function, great to see this!!

Reply:
this is just a validation of the BSP API is "correct" and the parameter is not only used in one case: CC3200 but in ESP32 too, probably this shows that simpler file systems need size prior open, ok, probably only I am happy about it :)

Reply:
I'm happy about it too! :) It would suck to download the entire size of the partition, only to have it fail at the end because there's not enough space. Glad our BSP is flexible enough, and glad the SDK devs keep it in mind.

Reply:
👍